src/transaction/types/script.rs

```diff
@@ -0,0 +1,700 @@
+use crate::transaction::{
+    checkable::{check_common_part, Checkable},
+    field::{
+        GasLimit, GasPrice, Inputs, Maturity, Outputs, ReceiptsRoot, Script as ScriptField,
+        ScriptData, Witnesses,
+    },
+    metadata::CommonMetadata,
+    Chargeable,
+};
+use crate::{CheckError, ConsensusParameters, Input, Output, TransactionRepr, Witness};
+use derivative::Derivative;
+use fuel_types::bytes::{SizedBytes, WORD_SIZE};
+use fuel_types::{bytes, Bytes32, Word};
+
+#[cfg(feature = "std")]
+use std::io;
+
+#[cfg(feature = "alloc")]
+use alloc::vec::Vec;
+use fuel_asm::Opcode;
+
+#[cfg(feature = "std")]
+use fuel_types::bytes::SerializableVec;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) struct ScriptMetadata {
+    pub common: CommonMetadata,
+    pub script_data_offset: usize,
+}
+
+#[derive(Debug, Clone, Derivative)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derivative(Eq, PartialEq, Hash)]
+pub struct Script {
+    pub(crate) discriminant: TransactionRepr,
```

Comment:
Why do we need to add the discriminant here? Since there's no valid case where this could be anything other than a script, shouldn't this be an associated constant?

Reply:
I added it as a field here https://github.com/FuelLabs/fuel-tx/pull/197/commits/3bfc13ae7c7ef52b051b21c9ce896edf388a4903 to highlight that it is part of the canonical serialization/deserialization to be more clear why `Create` and `Script` do it instead of `Transaction`

Reply:
My concern is that we are letting the desire for auto-derived serialization drive the structure of our core domain types in ways that may not be sound. Before we had compiler guarantees that a script tx would always have the correct discriminant, now it is a possible edge case to handle at runtime.

Reply:
Still doesn't feel right to have serialization concerns affecting our type design, but given it is private like you said I think we can defer this to a future tech-debt.

Reply:
Then let's remove it for now if it doesn't make thing cleaner=)